library/std/src/sys/unix/fs.rs

```diff
@@ -489,8 +489,18 @@ impl Iterator for ReadDir {
                     };
                 }
 
+                // Only d_reclen bytes of *entry_ptr are valid, so we can't just copy the
+                // whole thing (#93384).
+                let reclen = (*entry_ptr).d_reclen as usize;
+                let mut entry: dirent64 = mem::zeroed();
+                ptr::copy_nonoverlapping(
+                    entry_ptr as *const u8,
+                    &mut entry as *mut _ as *mut u8,
+                    reclen,
+                );
```

Comment:
I don't think we can copy `d_reclen` bytes either, since it might exceed the size of dirent structure:

> Note that while the call `fpathconf(fd, _PC_NAME_MAX)` returns the value 255 for most filesystems, on some filesystems  (e.g., CIFS, Windows SMB servers), the null-terminated filename that is (correctly) returned in d_name can actually exceed this size. In such cases, the d_reclen field will contain a value that exceeds the size of the glibc dirent structure shown above.

Reply:
Good point -- we would be better of destructuring the individual fields we need. This code is already allocating a separate copy of the full name.

Reply:
@tmiasko Oh right yeah, that was the whole reason for my initial patch.

@cuviper I'm happy to do that but it's a larger change, thought you might like this one first to unbreak CI.

Reply:
Ok, then maybe use `ptr::addr_of((*entry_ptr).d_name)` to figure out the offset, and copy the bytes up to that point. Or just copy those individual fields manually into the zeroed `entry`.

Reply:
Done!